### ISSUE TITLE
Unit test mocks

### ISSUE THREAD

**Issue by Sriep** [0chain/0chain#170]
https://github.com/0chain/0chain/pull/115, https://github.com/0chain/0chain/pull/87, https://github.com/0chain/0chain/pull/72, https://github.com/0chain/0chain/pull/61 and possibly https://github.com/0chain/0chain/pull/112 make use of unit test mocks. 

Currently, the code is mixed in with unit test in the above PRs, difficult to find and undocumented.

This issue is to discuss this potential new feature.

**Comment by Sriep** [comment]
My first impressions. I might add as I study the code.

1. Code for `mocks` should not be mixed up with production code. 
2. A new package, maybe called `mocks`, should be created to hold common mock code.
3. If necessary, structs in production can be turned into interfaces, and mock implementations put in the new mocks package.
4. If there is any generated code - not sure if there is - it should be generated by CI, probably as a merge check.

**Comment by Sriep** [comment]
[Here is an example from PR72 where mock code is being mixed with production code. ](https://github.com/0chain/0chain/pull/72/files?file-filters%5B%5D=.go#diff-b791befd3c5f2dbe8739faa8ae855e66c9bc40023e48d04574a60bde0f3ad20eR41). If we really need to be doing this, then I would suggest that global functions like `InitSession` be added to an interface object which can be implemented separately in the mock code.

**Comment by TheTeaParty** [comment]
So to not run around, the best practice suggests to have `mocks/` folder in the root of the project with all `mocks`. Some mocks are auto generated with tools like `mockery` some mocks we are forced to hand write. So in my experience it is good to have mocks/ folder for all mocks generated or not so testing code will be as simple as running `go test ./...` without requiring any additional .sh files or `go generate ./...`.

**Comment by Sriep** [comment]
So my suggested solution is to work on PR https://github.com/0chain/0chain/pull/61. I suggest splitting it up into PRs as follows:
1. go,mod changes https://github.com/0chain/0chain/pull/162
2. Any other production refactoring needed for `mocks`. 
3. New `mocks` package
4. Any `miner` bugs that need fixing.
5.  `miner` unit tests
6. The remaining PRs https://github.com/0chain/0chain/pull/72, https://github.com/0chain/0chain/pull/87 and https://github.com/0chain/0chain/pull/115 can be redone using the new mocks package. I don't think https://github.com/0chain/0chain/pull/112 really needs it.

**Comment by stewartie4** [comment]
Just to add my own thoughts to this:
A lot of the code is quite procedural and tightly coupled which makes mocking and unit testing in general quite difficult.
Going forward we could consider moving to small, distinct interface types which would be easily tested in isolation and trivial to replace with a mock during a unit test

**Comment by TheTeaParty** [comment]
> Just to add my own thoughts to this:
> A lot of the code is quite procedural and tightly coupled which makes mocking and unit testing in general quite difficult.
> Going forward we could consider moving to small, distinct interface types which would be easily tested in isolation and trivial to replace with a mock during a unit test

That one is for sure, but will need some time to reach that point.

**Comment by Sriep** [comment]
>A lot of the code is quite procedural and tightly coupled which makes mocking and unit testing in general quite difficult.

Difficult but entirely possible.

The advantage of a common mocks package is that they `speed up the writing of unit test` by crating common mocks that the developer does not need to re-write for each test.

Unsurprising, the main advantage is also the main disadvantage. The danger is that unit tests will `use code it does not need`, just because they are part of the mocks, or it's easier to start all tests with the same copy-paste. Possible results:
* Unit tests will take longer to run due to building up and tearing down structures they don't need.
* The actions of the method being tested will be obscured by code unneeded by the test method.
* By blindly reusing common mocks it's easy to miss some detail of how the test function operates.

The point of a unit test is to map inputs to outputs and confirm they match. The input and outputs should be made clear in some way to the reader.

Inputs.
1. Global variables such as [approvedMinters](https://github.com/0chain/0chain/blob/master/code/go/0chain.net/chaincore/chain/state/state_context.go#L13)
2. Input parameters. The obvious input parameter is [sometimes ignored](https://github.com/0chain/0chain/blob/master/code/go/0chain.net/smartcontract/minersc/fees.go#L302).
3. Hidden deep in some context object. A common example will be data stored in an MPT with [InsertTrieNode](https://github.com/0chain/0chain/blob/master/code/go/0chain.net/chaincore/chain/state/state_context.go#L41). These could be entered explicitly before each test, but there is a danger they will be hidden in a mocked object.
4. From config files. `0chain.yaml`, `sc.yaml`, `blobber.yaml` etc.  I think the config options used by the test function should be made obvious in some way. If the test fails for using a config option it should not be using, then that's a plus. It should be easy for the tester to run write tests with different config options. Does the function error correctly when `input > max_something`?

Outputs
1. Return value. Often ignored as 2. below is the main point of the tested function.
2. Hidden deep in various large objects. This includes MPTs as in 3. above, but anything returned by [StateContextI](https://github.com/0chain/0chain/blob/master/code/go/0chain.net/chaincore/chain/state/state_context.go#L31). In particular I have needed `GetTransfers` and `GetMints`.

**Comment by Sriep** [comment]
> A lot of the code is quite procedural and tightly coupled which makes mocking and unit testing in general quite difficult.

One effect of this is that it makes `magic number` tests of limited use. 
For instance inputs of `(2, 0.35, 2, {7,11,13})` give a result of `248.387096774` is not very informative. You would rather know the result equals something like `2 * 1000 * 0.35 * 11/(7+11+13)` and never know the actual value.

Hence I recommend the following format for unit tests, which I have been using.
```
                  |-> test function ---|
Input Parmeters --|                    |-> confirm expected = actual
                  |-> expected result -|
```
* Input Parameters: Sets up parameters for this test. Passes to `test function` and `expected  result` modules
* test function: Runs the `test function` plus sets up mocks. After running gathers together the various outputs and passes them to the `confirm expected module` Constrain all the messy details to this, so the other modules are as easy to read as possible.
* expected results: Only knows what results the test function should produce from any given input. Otherwise, 0chain agnostic.
* confirm expected = actual: Mostly a list of `require` statements that confirm the expected and actual outputs match for any given input.



You can now add as many `different tests` = `sets of input parameters` as you want, without changing the other three modules in any way.

**Comment by TheTeaParty** [comment]
@Sriep @guruhubb this is closed by #320
